File: Algorithms/1396. Design Underground System.rs
use std::collections::HashMap;

#[derive(Default)]
struct UndergroundSystem {
    time: HashMap<String, HashMap<String, (i32, i32)>>,
    customer: HashMap<i32, (String, i32)>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl UndergroundSystem {

    fn new() -> Self {
        UndergroundSystem::default()
    }
    
    fn check_in(&mut self, id: i32, station_name: String, t: i32) {
        self.customer.insert(id, (station_name, t));
    }
    
    fn check_out(&mut self, id: i32, station_name: String, t: i32) {
        let (start_station, start_t) = self.customer.remove(&id).expect("in");
        let (sum, count) = self
            .time
            .entry(start_station)
            .or_default()
            .entry(station_name)
            .or_default();
        *sum += t - start_t;
        *count += 1;
    }
    
    fn get_average_time(&mut self, start_station: String, end_station: String) -> f64 {
        let (sum, count) = self
            .time
            .entry(start_station)
            .or_default()
            .entry(end_station)
            .or_default();
        *sum as f64 / *count as f64
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * let obj = UndergroundSystem::new();
 * obj.check_in(id, stationName, t);
 * obj.check_out(id, stationName, t);
 * let ret_3: f64 = obj.get_average_time(startStation, endStation);
 */
 

File: Algorithms/1396. Design Underground System.ts
interface Trip {
  stationName: string;
  t: number;
}

interface AvgData {
  sum: number;
  count: number;
}

class UndergroundSystem {
  constructor() {
    this.customer = new Map();
    this.avarage = new Map();
  }

  private customer: Map<number, Trip>;
  private avarage: Map<string, AvgData>;

  checkIn(id: number, stationName: string, t: number): void {
    this.customer.set(id, { stationName, t });
  }

  checkOut(id: number, stationName: string, t: number): void {
    const checkIn = this.customer.get(id)!;
    const key = `${checkIn.stationName}-${stationName}`;
    const { sum, count } = this.avarage.get(key) ?? { sum: 0, count: 0 };
    this.avarage.set(key, { sum: sum + (t - checkIn.t), count: count + 1 });
  }

  getAverageTime(startStation: string, endStation: string): number {
    const { sum, count } = this.avarage.get(
      `${startStation}-${endStation}`
    ) ?? { sum: 0, count: 0 };
    return count !== 0 ? sum / count : 0;
  }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * var obj = new UndergroundSystem()
 * obj.checkIn(id,stationName,t)
 * obj.checkOut(id,stationName,t)
 * var param_3 = obj.getAverageTime(startStation,endStation)
 */
